packages/style-engine/class-wp-style-engine.php

```diff
@@ -66,106 +223,213 @@ public static function get_instance() {
-					}
+			// In the future there might be a flag in the option to output
+			// inline CSS rules (for HTML style attributes) vs selectors + rules for style tags.
+			foreach ( $css_rules as $rule => $value ) {
+				$filtered_css = esc_html( safecss_filter_attr( "{$rule}: {$value}" ) );
+				if ( ! empty( $filtered_css ) ) {
+					$css_output .= $filtered_css . '; ';
 				}
 			}
 		}
 
-		return $output;
+		if ( ! empty( $css_output ) ) {
+			$styles_output['css'] = trim( $css_output );
+		}
+
+		if ( ! empty( $classnames ) ) {
+			$styles_output['classnames'] = implode( ' ', array_unique( $classnames ) );
+		}
+
+		return $styles_output;
 	}
 
 	/**
-	 * Returns a CSS ruleset for box model styles such as margins, padding, and borders.
+	 * Default style value parser that returns a CSS ruleset.
+	 * If the input contains an array, it will be treated like a box model
+	 * for styles with sides such as margins, padding, and borders.
 	 *
-	 * @param string|array $style_value    A single raw Gutenberg style attributes value for a CSS property.
-	 * @param string       $style_property The CSS property for which we're creating a rule.
+	 * @param string|array $style_value      A single raw Gutenberg style attributes value for a CSS property.
+	 * @param array        $style_definition A single style definition from BLOCK_STYLE_DEFINITIONS_METADATA.
 	 *
 	 * @return array The class name for the added style.
 	 */
-	public static function get_css_box_rules( $style_value, $style_property ) {
+	protected static function get_css_rules( $style_value, $style_definition ) {
 		$rules = array();
 
 		if ( ! $style_value ) {
 			return $rules;
 		}
 
+		$style_properties = $style_definition['properties'];
+
+		// We assume box model-like properties.
 		if ( is_array( $style_value ) ) {
 			foreach ( $style_value as $key => $value ) {
-				$rules[ "$style_property-$key" ] = $value;
+				$side_property           = strtr( $style_properties['sides'], array( '$side' => _wp_to_kebab_case( $key ) ) );
+				$rules[ $side_property ] = $value;
 			}
 		} else {
-			$rules[ $style_property ] = $style_value;
+			$rules[ $style_properties['default'] ] = $style_value;
+		}
+
+		return $rules;
+	}
+
+	/**
+	 * Style value parser that returns a CSS ruleset for style groups that have 'top', 'right', 'bottom', 'left' keys.
+	 * E.g., `border.top{color|width|style}.
+	 *
+	 * @param array $style_value      A single raw Gutenberg style attributes value for a CSS property.
+	 * @param array $style_definition A single style definition from BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @return array The class name for the added style.
+	 */
+	protected static function get_css_side_rules( $style_value, $style_definition ) {
+		$rules = array();
+
+		if ( ! is_array( $style_value ) || empty( $style_value ) ) {
+			return $rules;
 		}
+
+		foreach ( $style_value as $key => $value ) {
+			$side_style_definition_path = array( $style_definition['path'][0], $key );
```

Comment:
I've made some assumptions both here `$style_definition['path'][0]` and below `$style_definition['path'][1]`.

`$style_definition['path'][0]` assumes the path array's first value is the style property, e.g., `border`. It also implies that `border` contains the necessary definitions (color, width, etc) for the sides.

The use of `$style_definition['path'][1]` assumes the path array's final item corresponds to the side name.

Both of these are fine in my opinion because nothing will work if the paths are not right, but it does seems a bit "clever" and unreadable.

A possibility is to add explicit properties to the style's metadata such as `inherits` or something. 🤷 

The latter

Reply:
Good question about the assumptions here — just to walk through how I understand it:

* `$style_definition['path'][0]` will map to the parent level property that we're looking up (in this case `border` but could theoretically be any of the other root level props of the metadata).
* `$style_definition['path'][1]` will map to the slug for the side.

It took me a little while to read through and work out what was going on here, but I think the assumptions are good! What I'm wondering is whether we can rename things slightly to improve readability.

In the function signature the `$style_definition` prop will always be the definition for the side (at least based on the props that have opted in to use this callback). Would it improve clarity to rename this prop to `$side_style_definition`?

Then, where we have the existing `$side_style_definition`, what if we renamed this to `$style_definition`.

We could also store the `['path'][0]` and `['path'][1]` values in variables, so they've got a concrete name for what they're referring to. Putting it all together, it might look like:

```php
	protected static function get_css_side_rules( $style_value, $side_style_definition ) {
		$rules = array();

		if ( ! is_array( $style_value ) || empty( $style_value ) || empty( $side_style_definition['path'] ) ) {
			return $rules;
		}

		// The first item in the style definition path array tells us the style property, e.g., "border".
		// We use this to get a corresponding CSS style definition such as "color" or "width" from the same group.
		$definition_group_key = $side_style_definition['path'][0];
		$side                 = $side_style_definition['path'][1];

		foreach ( $style_value as $css_property => $value ) {
			if ( empty( $value ) ) {
				continue;
			}

			$style_definition_path = array( $definition_group_key, $css_property );
			$style_definition      = _wp_array_get( self::BLOCK_STYLE_DEFINITIONS_METADATA, $style_definition_path, null );

			if ( $style_definition && isset( $style_definition['properties']['sides'] ) ) {
				// Set a CSS var if there is a valid preset value.
				$slug = isset( $side_style_definition['css_vars'][ $css_property ] ) ? static::get_slug_from_preset_value( $value, $css_property ) : null;
				if ( $slug ) {
					$css_var = strtr(
						$side_style_definition['css_vars'][ $css_property ],
						array( '$slug' => $slug )
					);
					$value   = "var($css_var)";
				}
				// The second item in the style definition path array refers to the side property, e.g., "top".
				$side_css_property           = strtr( $style_definition['properties']['sides'], array( '$side' => $side ) );
				$rules[ $side_css_property ] = $value;
			}
		}
		return $rules;
	}
```

What do you think? It's probably a fairly small change, so I'm not sure if that improves things very much, so feel free to skip it — but typing some of that out helped me understand what's going on a bit better 😀. At the very least, this is internal behaviour to the style engine, so I think it's reasonably safe for us to go with the general approach here, and we can always refine it in follow-ups if need be, without affecting the exposed API of the engine.

Reply:
This is much better thank you. If it helps folks who are reading it for the first time it's definitely worth it.